lib/iteration.rb

```diff
@@ -0,0 +1,109 @@
+def benchmark(num = 1)
+  start = Time.now
+  num.times { yield }
+  elapsed_time = (Time.now - start) / num
+end
+
+def race(procs, num = 1)
+  arr = []
+  procs.each do |proc|
+    runtime = benchmark(num)
+    arr << runtime
+  end
+  index = arr.index(arr.max)
+  procs[index]
+end
+
+def eval_block(*args, &prc)
+  if prc.nil?
+    "NO BLOCK GIVEN"
+  else
+    p yield *args
+  end
+end
+
+def factors(num)
+  (1..num).select { |n| num % n == 0 }
+end
+
+class Array
+  def bubble_sort!(&prc)
+    sorted = false
+    until sorted
+      sorted = true
+      (0..self.length - 2).each do |idx|
```

Comment:
Because in an array like [ 1, 3, 4, 5, 6 ] which has 5 elements, I only want to sort from position 0-3, which is basically the numbers 1, 3, 4, 5. Once I compare 5 with 6, I don't need to move forward anymore because I don't have any thing to compare the last element with.